[MVP1][tk116] Alteração do módulo catalogmanager.persistence

patymori opened the pull request:
Fixes #116.

- Criação de `insert_file` nos DBManagers
- Criação de `ChangesService.register` com o novo conteúdo do registro
de mudança
- Criação de `DatabaseService.add_file` para persistir arquivos usando o
DBManager e registrando a mudança com `ChangesService.register`
- Ajustes nos fixtures para executar os testes sem duplicação
- Testes unitários e de integração do módulo persistence para o novo
registro de artigos

Commit:
Alteração do módulo catalogmanager.persistence

- Criação de `insert_file` nos DBManagers
- Criação de `ChangesService.register` com o novo conteúdo do registro
de mudança
- Criação de `DatabaseService.add_file` para persistir arquivos usando o
DBManager e registrando a mudança com `ChangesService.register`
- Ajustes nos fixtures para executar os testes sem duplicação
- Testes unitários e de integração do módulo persistence para o novo
registro de artigos

Commit:
Merge branch 'master' of github.com:patymori/catalogmanager into mvp1-tk106

Commit:
Adding Prometheus metric to add_file

Commit:
DBManager.insert_file returns absolut file URL

robertatakenaka commented on persistence/services.py:
```diff
@@ -78,6 +78,20 @@ def register_change(self,
         return change_record['record_id']
 
+    def register(self, record_id, change_type):
+        sequencial = str(self.seqnum_generator.new())
```
@patymori : sendo `str`, não corre o risco de a ordem válida ser 1, 10, 11 no lugar de 1, 2, 3?

gustavofonseca replied:
Essa dúvida deveria, idealmente, ser respondida por meio de testes de unidade ;)

robertatakenaka commented on persistence/services.py:
```diff
@@ -253,6 +267,19 @@ def get_attachment_properties(self, document_id, file_id):
+        file_id: ID do arquivo
+        content: conteúdo do arquivo
+        """
+        self.db_manager.insert_file(file_id=file_id, content=content)
```
@patymori : está sendo considerado que tanto os arquivos e os manifestos serão gravados usando o mesmo db_manager. Isso será disvinculado futuramente, correto?

patymori replied:
@robertatakenaka, isso foi um problema q eu decidi abordar de outra forma: estou pensando em gerenciar 2 db_managers diferentes, 1 para cada database

robertatakenaka replied:
@patymori ok, minha dúvida é 2 db_service (1 para cada db_manager) ou 1 db_service com 2 db_managers?

patymori replied:
Desculpe-me, @robertatakenaka, 2 db_services.

robertatakenaka commented on persistence/tests/conftest.py:
```diff
@@ -76,47 +70,79 @@ def fin():
-        db_service.db_manager.drop_database()
-        db_service.changes_service.changes_db_manager.drop_database()
+        try:
+            db_service.db_manager.drop_database()
```
@patymori : db_service deveria ter um drop_database para apagar em cascata.

robertatakenaka commented on persistence/tests/test_changes.py:
```diff
@@ -13,22 +13,39 @@ def get_article_record(content={'Test': 'ChangeRecord'}):
 
 
-def test_register_create_change(database_service):
+def test_register_change_create(database_service):
```
@patymori : sim fica melhor invertido

robertatakenaka commented on persistence/tests/test_changes.py:
```diff
@@ -13,22 +13,39 @@ def get_article_record(content={'Test': 'ChangeRecord'}):
+    )
+    assert change_id is not None
+    check_change = dict(
+        database_service.changes_service.changes_db_manager.database[change_id]
```
Por esta linha, me parece que deveríamos ter previsto um changes_service.get(change_id)

patymori replied:
Penso que não é necessário pois não faz sentido para a solução, visto que a aplicação não oferecerá consulta/leitura de uma mudança pelo ID, certo?

robertatakenaka replied:
ok

robertatakenaka commented on persistence/tests/test_databases.py:
```diff
@@ -489,3 +489,54 @@ def compare_documents(document, expected):
+        file_id='href_file1',
+        content=xml_test.encode('utf-8')
+    )
+    file = db_manager_test.database.get(
```
por que a recuperação do file não é `db_manager_test.get_file(file_id)`?

patymori replied:
Porque eu tava implementando só o incluir. Não existe ainda nessa implementação um recuperar arquivo pelo ID. Podemos refatorar esse teste quando ele existir, o que acha?

robertatakenaka replied:
sim

Commit:
Ajuste do sequencial na inserção do novo registro de mudança

robertatakenaka commented on persistence/tests/test_changes.py:
```diff
@@ -13,22 +14,67 @@ def get_article_record(content={'Test': 'ChangeRecord'}):
+        changes_list[i]['change_id'] < changes_list[i + 1]['change_id']
+        for i in range(len(changes_list) - 1)
+    ])
+
```
@patymori : Referente às linhas 72 a 75:
Pode-se comparar também usando ```assert sorted(items) == items```

Por curiosidade, fiz um teste de desempenho.

```assert sorted(_items) == _items``` é mais eficiente. 

Para 10 items, a diferença é irrelevante: 1s.
Para 100 items, a diferença é 66 - 38.
```
>>> t1 = datetime.now(); _items = [item['x'] for item in items];assert sorted(_items) == _items;t2 = datetime.now(); t = t2-t1;print(t); print(t1);print(t2)
0:00:00.000038
2018-06-18 08:41:37.630788
2018-06-18 08:41:37.630826

>>> t1 = datetime.now(); assert all([items[i]['x'] < items[i+1]['x'] for i in range(len(items)-1)]);t2 = datetime.now(); t = t2-t1;print(t); print(t1);print(t2)
0:00:00.000066
2018-06-18 08:41:51.818453
2018-06-18 08:41:51.818519
```

Commit:
Alteração da camada de persistência e ajustes do WEB App

- Ajustes em `managers` para as mudanças em `persistence`.
- Alteração do ArticleDocument para prever o registro de versões
diferentes do artigo.
- Ajuste do ArticleManager para as alterações do ArticleDocument, o
gerenciamento de um DBManager para os arquivos brutos, e a implementação
do método `add_document`.
Ajuste do método post na view de artigo.
- Criação de teste funcional para a funcionalidade de registro de novo
artigo.

gustavofonseca commented:
@patymori, você poderia resolver os conflitos?

gustavofonseca commented on managers/__init__.py:
```diff
@@ -1,6 +1,6 @@
 from managers.article_manager import ArticleManager
 from managers.exceptions import ManagerFileError
-from managers.models.article_model import ArticleDocument
+from managers.models.article_model import ArticleDocument, InvalidXMLContent
```
_InvalidXMLContent_ é uma exceção. Não deveria estar baixo _managers.exceptions_?

patymori replied:
Criei `managers.exceptions` para ter somente as exceções que serão externalizadas, com o intensão de manter uma interface do que seria o módulo de persistência da aplicação WEB. Não sei se faz sentido.

gustavofonseca commented on managers/__init__.py:
```diff
@@ -35,9 +35,12 @@ def _get_article_manager(**db_settings):
     database_config = db_settings
```
aqui _database_config_ não serve para nada, uma vez que as cópias das linhas 36 e 38 poderiam partir diretamente de _db_settings_

gustavofonseca commented on managers/__init__.py:
```diff
@@ -57,11 +60,33 @@ def create_file(filename, content):
+        codificado em XML
+    :rtype: str
+    """
+    try:
```
Sugiro envolver na estrutura _try..except_ o mínimo de código possível, i.e., apenas o que de fato tem o potencial de lançar a exceção que será tratada. Nessa função seria algo como:

```
article_manager = _get_article_manager(**db_settings)
article_document = ArticleDocument(article_id)
try:
    article_document.add_version(xml_id, xml_file)
except  InvalidXMLContent as e:
    raise ManagerFileError(message=e.message) from e
else:
    return article_manager.add_document(article_document)
```

gustavofonseca commented on managers/__init__.py:
```diff
@@ -165,7 +190,8 @@ def set_assets_public_url(article_id, xml_content, assets_filenames,
     """
     xml_file = File(file_name="xml_file.xml", content=xml_content)
-    article = ArticleDocument(article_id, xml_file)
+    article = ArticleDocument(article_id)
```
Cheiro de inicialização incompleta de _ArticleDocument_

patymori replied:
_ArticleDocument_ pode ser instanciado sem o conteúdo do XML, como ao recuperar o manifesto. Há a possibilidade de colocá-lo como um parâmetro não obrigatório. Melhor?

gustavofonseca commented on managers/__init__.py:
```diff
@@ -165,7 +190,8 @@ def set_assets_public_url(article_id, xml_content, assets_filenames,
+    article = ArticleDocument(article_id)
+    article.xml_file = xml_file
     for name in article.assets:
         if name in assets_filenames:
```
Não entendi a necessidade do argumento _assets_filenames_. O nome de identificação dos ativos digitais contigos no XML não podem ser obtidos inspecionando o próprio XML?

patymori replied:
Esse método só foi alterado por conta da alteração de _ArticleDocument_ e vem de implementações antigas que, provavelmente, serão alteradas. Podemos criar algum comentário nesse ponto para lembrar de revisá-lo ou criamos uma issue de melhoria?

gustavofonseca commented on managers/models/article_model.py:
```diff
@@ -34,20 +45,34 @@ def href(self, value):
+        self.xml_tree = ArticleXMLTree(xml_content)
+        if self.xml_tree.xml_error:
+            raise InvalidXMLContent
+        checksum = hashlib.sha1(xml_content).hexdigest()
```
_checksum_ deveria ser um método de _ArticleXMLTree_

gustavofonseca commented on managers/article_manager.py:
```diff
@@ -27,17 +27,21 @@ class ArticleManagerMissingAssetFileException(Exception):
 
     def receive_xml_file(self, id, xml_file):
-        article = ArticleDocument(id, xml_file)
+        article = ArticleDocument(id)
```
Inicialização incompleta de _ArticleDocument_?

gustavofonseca reviewed (commented):
Estes são comentários de uma revisão parcial. Vou continuar amanhã. Prometo ;)

Em geral eu diria que a implementação está boa. Só gostaria de chamar à atenção para o excesso de responsabilidades de alguns métodos como o ArticleDocument.add_version por exemplo.

Commit:
Merge branch 'master' of github.com:patymori/catalogmanager into mvp1-tk106

Commit:
Ajustes relacionados aos comentários do PR

- Ajuste do `managers._get_article_manager`
- Ajuste em `post_article` no tratamento de exceção e nos testes
- Criada property `ArticleXMLTree.checksum`
